### ISSUE TITLE
Does not build against librealsense v2.16.0

### ISSUE THREAD

**Issue by alexswerner** [realsenseai/realsense-ros#451]
I tried to build this and librealsense as ROS packages with catkin on Ubuntu xenial and ROS kinetic. Against v2.15.0 it builds successfully but against v2.16.0 it fails. The error is:
<pre>
In file included from /home/workspace/devel/include/librealsense2/hpp/rs_device.hpp:8:0,
                 from /home/workspace/devel/include/librealsense2/hpp/rs_record_playback.hpp:8,
                 from /home/workspace/devel/include/librealsense2/hpp/rs_context.hpp:8,
                 from /home/workspace/devel/include/librealsense2/rs.hpp:9,
                 from /home/workspace/src/realsense/realsense2_camera/src/../include/../include/realsense_node_factory.h:11,
                 from /home/workspace/src/realsense/realsense2_camera/src/../include/base_realsense_node.h:6,
                 from /home/workspace/src/realsense/realsense2_camera/src/base_realsense_node.cpp:1:
/home/workspace/devel/include/librealsense2/hpp/rs_sensor.hpp: In instantiation of 'void rs2::frame_callback<T>::on_frame(rs2_frame*) [with T = rs2::asynchronous_syncer; rs2_frame = rs2_frame]':
/home/workspace/src/realsense/realsense2_camera/src/base_realsense_node.cpp:1396:1:   required from here
/home/workspace/devel/include/librealsense2/hpp/rs_sensor.hpp:108:13: error: no match for call to '(rs2::asynchronous_syncer) (rs2::frame)'
             on_frame_function(frame{ fref });
             ^
</pre>

When i run git bisect the first offending commit is 6850677e0473bd01a2078ef6e0398a617539cf56 however
with a different error message:
<pre>
/home/workspace/src/librealsense/src/../include/librealsense2/hpp/rs_frame.hpp: In instantiation of 'void rs2::frameset::foreach(T) const [w
ith T = librealsense::align::should_process(const rs2::frame&)::<lambda(rs2::frame&)>]':
/home/workspace/src/librealsense/src/proc/align.cpp:572:134:   required from here
/home/workspace/src/librealsense/src/../include/librealsense2/hpp/rs_frame.hpp:965:23: error: no match for call to '(librealsense::align::sh
ould_process(const rs2::frame&)::<lambda(rs2::frame&)>) (rs2::frame)'
                 action(frame(fref));
</pre>

I was successful locating the original api change which causes the build to fail. Staying with librealsense v2.15.0 is not an option as it is not compatible with kernels >=4.16.

Let me know if I can provide more information.

**Comment by atyshka** [comment]
Also having this error

**Comment by timkambic** [comment]
Same problem here

**Comment by nindanaoto** [comment]
same problem

**Comment by Jaeyoung-Lim** [comment]
Same problem.

Downgrading the librealsense to 2.14 helped

**Comment by ChainBreak** [comment]
Same problem. Making me sad

**Comment by robotlinker** [comment]
I have the same problem
![screenshot from 2018-08-30 02-26-36](https://user-images.githubusercontent.com/24693929/44782234-24d46980-abb9-11e8-9cf5-6be7b4eef9cf.png)

**Comment by robotlinker** [comment]
same problem with latest SKD

**Comment by doronhi** [comment]
Fixed. Checkout latest release, Pull request #452.

**Comment by nindanaoto** [comment]
The build is now passing with latest commit against librealsense v2.16.0.
Thank you.

**Comment by itabr** [comment]
I am still getting the same problem with latest Intel® RealSense™ SDK 2.0. I tried building from source and got the same error.

**Comment by bhaskara** [comment]
I get a compile error about `invoke` not being found when I try to build this after PR #452 .  Suggestions?

**Comment by nindanaoto** [comment]
I used the librealsense provided by ubuntu repository for 18.04. Not the one builded from the source.

**Comment by bhaskara** [comment]
I'm on 16.04 + ROS Kinetic, using the ubuntu librealsense2 2.15.  Is this a supported configuration @doronhi ?

**Comment by doronhi** [comment]
@bhaskara, The latest update on intel-ros is checked with librealsense2 2.**16**. The same configuration is as you said.

**Comment by doronhi** [comment]
@itabr , Is it possible that you still have remnants of previous versions on your machine? Can you confirm that you get the same error on a clean machine, using the latest commit of intel-ros and librealsense SDK 2.0 (build 2.16.0)? If so, can you provide additional data, such as a build log, what version of OS you are using, etc.?

**Comment by bhaskara** [comment]
Thanks @doronhi .  My specific error was that it couldn't find the function `invoke` from your patch.  Do you know where that's from?  Is it a librealsense thing?

**Comment by itabr** [comment]
Thanks @doronhi I can confirm that intel-ros/realsense with the latest master commit works with IntelRealSense/librealsense v2.16.0 (Installed from [Debian Package](https://github.com/IntelRealSense/librealsense/blob/master/doc/distribution_linux.md#installing-the-packages)) and Ubuntu 16.04.

**Comment by doronhi** [comment]
@bhaskara, invoke is indeed a librealsense thing...
It belongs to class asynchronous_syncer : public processing_block (defined in file: include\librealsense2\hpp\rs_processing.hpp)

**Comment by bhaskara** [comment]
Ah, thanks for the pointer.  Looking at the source, in 2.15, `asynchronous_syncer` does *not* inherit from `processing_block`, and thus `PipelineSyncer` does not inherit `invoke` (https://github.com/IntelRealSense/librealsense/blob/1e8aa8002bc3d6afbfc7b733ea4a19864bbeca14/include/librealsense2/hpp/rs_processing.hpp#L330v).  It looks, therefore, like upgrading to the above patch requires simultaneously upgrading to 2.16.  

It might be convenient to have a version of the patch that allowed 2.15 (e.g., in my setup the choice of librealsense2 version and the choice of intel-ros version happen in two different parts of the provisioning and would be nice to integration test independently).    I tried just copying over the definition of invoke into `base_realsense_node.h`, but got further errors (e.g., it relies on the fact that `process_block` is a friend of `frame`) that I haven't resolved yet.

Let me know if any of this seems way off base.

**Comment by doronhi** [comment]
Hi @bhaskara.
In general I think it's advisable to avoid multiple versions, each of which needs maintenance. As realsense SDK 2.16 is not expected to have any of the features of 2.15 not supported, there doesn't seem to be a reason why not to upgrade to 2.16 and therefor no reason to maintain another version of intel-ros supporting 2.15.
I do think it would be very helpful if a proper error message was to be displayed while building or installing intel-ros, after automatically checking the version of librealsense and finding it inappropriate. This, however, could take me a while as this is not very high in the priority order.

**Comment by yeyanlei** [comment]
![realsense_problem](https://user-images.githubusercontent.com/36154470/45080207-f9a4c980-b126-11e8-802a-8e919d20f3c0.gif)
@doronhi , I meet the same problem( error: no match for call to ‘(rs2::asynchronous_syncer) (rs2::frame)’). could you give me some suggestion. Thanks for your help

**Comment by Achllle** [comment]
#452 made building fail with librealsense 2.15 indeed. Updating librealsense and rebuiling along with this update made it work again.

**Comment by AndyZe** [comment]
I am getting the `invoke() was not declared` error as well. I installed librealsense by binary (according to this page: https://github.com/IntelRealSense/librealsense/blob/master/doc/distribution_linux.md#installing-the-packages). The git commit of the realsense package I cloned is 6f2c7c8  (It fails on one computer. On a different computer, it's fine). `realsense_viewer` works fine in both cases.

Manually updating to librealsense 2.16 did fix it, as others have said (although more errors were encountered along the way...)
